Separate config section for each db engine

mattias-p opened the pull request:
## Purpose

This PR is about having separate a separate config section for each db engine.


## Context

Fixes #719.

As @matsduf pointed out, PR #717 must be merged before this is merged or else there is a risk of filling up the log files.


## Changes

Config format:
* New sections are added. The new section names are upper-case-only for consistency with previous sections.
* The old properties are deprecated.

Implementation:
* New accessor methods for the properties are added to Config.pm. They're named after the config section and property name, with an underscore in between. There is some precedence for this naming scheme in Config.pm.
* If the new properties are absent in the config file, the old ones are used as fallbacks to make sure old config files keep working.
* The accessors for the old properties are removed.
* The DB_connection_string() method didn't really belong in the Config module so I inlined it into the DB modules.

Installation:
* Newly obsolete steps in the installation instruction are cleaned up.
* The included config file is updated to the new format.

CI:
* The backend config files for travis are updated to the new format.

Database initialization:
* The initialization scripts now use the adapter class - they used to read accessors on the config object and construct its own handle. (Note that the SQLite initialization script is only used in the installation instruction while the MySQL and PostgreSQL dittos are only used for CI.)

Database upgrade:
* The database upgrade scripts are updated the same way as the database initialization scripts.

Logging:
* It's now possible to override ZM_BACKEND_RPCAPI_LOGLEVEL and ZM_BACKEND_TESTAGENT_LOGLEVEL when starting the daemons using the supplied init scripts.
* Logging of database connections is now more informative.
* Log messages emitted during the loading for the config file during startup is now output to the terminal (instead of being silently ignored).

## How to test this PR

This is a new feature, so older versions of Zonemaster need not be tested.

For more convenient testing of the database settings, set ZM_BACKEND_RPCAPI_LOGLEVEL and ZM_BACKEND_TESTAGENT_LOGLEVEL to notice or lower.

For each database engine:
1. Get the source from this PR.
   ```sh
   git clone --origin mattias-p --branch 719-db-sections https://github.com/mattias-p/zonemaster-backend.git
   cd zonemaster-backend
   ```
2. Verify that both daemons can be installed and that the SQLite database initialization works.
   1. Install Backend for according to the installation instruction, but instead of `cpanm Zonemaster::Backend` run `cpanm .`.
3. Verify that the installation uses the new configuration format.
      ```sh
      cat /etc/zonemaster/backend_config.ini
      ```
4. Verify that both daemons can be started using the old configuration format.
   1. Copy the share/backend_config.ini from the current develop branch into /etc/zonemaster.
      ```sh
      git remote add origin https://github.com/zonemaster/zonemaster-backend.git
      git fetch origin
      git checkout origin/develop
      sudo install -v -m 640 -g zonemaster share/backend_config.ini /etc/zonemaster/
      ```
   2. Update the DB.engine property for the current database engine according to the installation instruction.
   3. Stop the daemons.
      ```sh
      sudo service stop zm-rpcapi
      sudo sercive status zm-rpcapi
      sudo service stop zm-testagent
      sudo sercive status zm-testagent
      ```
   4. Start the daemons.
      ```sh
      sudo service start zm-rpcapi
      sudo sercive status zm-rpcapi
      sudo service start zm-testagent
      sudo sercive status zm-testagent
      ```
5. Verify that deprecation warnings are emitted to the log files of both rpcapi and testagent if the old configurations properties are present when the configuration file is loaded.
   ```sh
   cat /var/log/zonemaster/zm-testagent.log | grep deprecated
   journalctl -u zm-rpcapi.service | grep deprecated
   ```

matsduf commented on docs/Configuration.md:
```diff
@@ -31,32 +30,118 @@ SQLite            | `SQLite`
 The name of the user with sufficient permission to access the database.
 
 Ignored by the SQLite database engine.
 
```
I think the four lines above can be removed to keep it shorter. The references to the new configuration makes it very clear.

mattias-p replied:
I like to keep the documentation of deprecated APIs even if we provide a migration path away from them. But sure, I'll remove it this time. I think it's a good thing that we stop and think about how we deprecate things.

matsduf commented on lib/Zonemaster/Backend/Config.pm:
```diff
@@ -37,6 +37,20 @@ sub load_config {
+    }
+    if ( defined $self->{cfg}->val( 'DB', 'database_name' ) ) {
+        $log->warning( "Use of deprecated config property DB.database_name. Use MYSQL.database, POSTGRESQL.database or SQLITE.file instead." );
+    }
```
PR #717 must be merged before this is merged or else there is a risk of filling up the log files.

mattias-p replied:
I included this comment in the PR description.

matsduf commented:
This looks good.

pnax commented on docs/Configuration.md:
```diff
@@ -31,32 +30,103 @@ SQLite            | `SQLite`
+**Deprecated.** Use [MYSQL.user] or [POSTGRESQL.user] instead.
 
-Ignored by the SQLite database engine.
+The [MYSQL.user] and [POSTGRESQL.user] properties take precedense over this.
```
`precedense` -> `precedence`

pnax commented on docs/Configuration.md:
```diff
@@ -31,32 +30,103 @@ SQLite            | `SQLite`
+**Deprecated.** Use [MYSQL.password] or [POSTGRESQL.password] instead.
 
-Ignored by the SQLite database engine.
+The [MYSQL.password] and [POSTGRESQL.password] properties take precedense over this.
```
`precedense` -> `precedence`

pnax commented on docs/Configuration.md:
```diff
@@ -31,32 +30,103 @@ SQLite            | `SQLite`
+**Deprecated.** Use [MYSQL.host] or [POSTGRESQL.host] instead.
 
-Ignored by the SQLite database engine.
+The [MYSQL.host] and [POSTGRESQL.host] properties take precedense over this.
```
`precedense` -> `precedence`

pnax commented on docs/Configuration.md:
```diff
@@ -31,32 +30,103 @@ SQLite            | `SQLite`
-it holds the full path to the SQLite database file.
+**Deprecated.** Use [MYSQL.database], [POSTGRESQL.database] or [SQLite.file] instead.
+
+The [MYSQL.database], [POSTGRESQL.database], [SQLite.file] properties take precedense
```
`precedense` -> `precedence`

pnax commented on docs/Configuration.md:
```diff
@@ -31,32 +30,103 @@ SQLite            | `SQLite`
+
+### host
+
+The host name of the machine on which the MYSQL server is running.
```
This one can have a lowercase `y`, `MYSQL` -> `MySQL` ;-)

pnax commented on docs/Configuration.md:
```diff
@@ -205,15 +275,28 @@ zero minutes, then the default value (600 seconds) is used.
 --------
 
-[Installation instructions]:          Installation.md
+[DB.database_host]:                   #db-section
+[DB.database_name]:                   #db-section
+[DB.password]:                        #db-section
+[DB.user]:                            #db-section
```
Why not pointing to the right subsection directly ?

mattias-p replied:
I was thinking that the password and user headings were ambiguous and I wanted to keep things consistent for the others too. I'd forgotten that Github has a way of disambiguating the headings.

Thanks for pointing this out!

pnax replied:
Yes I totally agree that this is not the easiest to maintain.

pnax commented on docs/Configuration.md:
```diff
@@ -205,15 +275,28 @@ zero minutes, then the default value (600 seconds) is used.
 [ISO 639-1]:                          https://en.wikipedia.org/wiki/ISO_639-1
+[Installation instructions]:          Installation.md
+[Language tag]:                       API.md#language-tag
+[MYSQL.database]:                     #mysql-section
+[MYSQL.host]:                         #mysql-section
+[MYSQL.password]:                     #mysql-section
+[MYSQL.user]:                         #mysql-section
+[POSTGRESQL.database]:                #postgresql-section
+[POSTGRESQL.host]:                    #postgresql-section
+[POSTGRESQL.password]:                #postgresql-section
+[POSTGRESQL.user]:                    #postgresql-section
```
same

pnax commented on docs/Configuration.md:
```diff
@@ -205,15 +275,28 @@ zero minutes, then the default value (600 seconds) is used.
 [Profile JSON files]:                 https://github.com/zonemaster/zonemaster-engine/blob/master/docs/Profiles.md
 [Profile names]:                      API.md#profile-name
 [Profiles]:                           Architecture.md#profile
+[SQLite.file]:                        #sqlite-section
```
same

pnax commented on lib/Zonemaster/Backend/Config.pm:
```diff
@@ -58,22 +72,130 @@ sub BackendDBType {
-    my ($self) = @_;
+=head2 MYSQL_database
+
+Returns the L<MYSQL.database|https://github.com/zonemaster/zonemaster-backend/blob/master/docs/Configuration.md#mysql>
```
The fragment should be `#mysql-section`, it is even possible to point directly to the right section with `#database`

mattias-p replied:
It's quite error prone to link to the auto-suffixed anchors. But let's try it out. If it's too painful we'll go back to just linking to the headings.

pnax replied:
Ok, we'll see how it goes, to be honest I don't have a strong opinion on this. I just wanted to know if you made this choice on purpose and I now totally understand your first move. Now that you updated it, let's go for a try ;)

pnax commented on lib/Zonemaster/Backend/Config.pm:
```diff
@@ -58,22 +72,130 @@ sub BackendDBType {
+=head2 MYSQL_database
+
+Returns the L<MYSQL.database|https://github.com/zonemaster/zonemaster-backend/blob/master/docs/Configuration.md#mysql>
+property from the loaded config, or the L<DB.database_name|https://github.com/zonemaster/zonemaster-backend/blob/master/docs/Configuration.md#db>
```
The fragments should be `#db-section`, it is even possible to point directly to the right section with `#database_name`

pnax commented on lib/Zonemaster/Backend/Config.pm:
```diff
@@ -58,22 +72,130 @@ sub BackendDBType {
+sub MYSQL_database {
+    my ( $self ) = @_;
+
+    return $self->{cfg}->val( 'MYSQL', 'database' ) // $self->{cfg}->val( 'DB', 'database_name' );
```
Will it fail if `DB.database_name` is unspecified ?

mattias-p replied:
It'll return undef, and the caller will most likely fail at some point. But it's like that for all the properties. I plan to take care of this in #685.

pnax commented on lib/Zonemaster/Backend/Config.pm:
```diff
@@ -58,22 +72,130 @@ sub BackendDBType {
-    my ($self) = @_;
+=head2 MYSQL_password
+
+Returns the L<MYSQL.password|https://github.com/zonemaster/zonemaster-backend/blob/master/docs/Configuration.md#mysql>
+property from the loaded config, or the L<DB.password|https://github.com/zonemaster/zonemaster-backend/blob/master/docs/Configuration.md#db>
```
`#mysql-section` or `#password-1`
`#db-section` or `#password`

pnax commented on lib/Zonemaster/Backend/Config.pm:
```diff
@@ -58,22 +72,130 @@ sub BackendDBType {
+
+=head2 MYSQL_user
+
+Returns the L<MYSQL.user|https://github.com/zonemaster/zonemaster-backend/blob/master/docs/Configuration.md#mysql>
+property from the loaded config, or the L<DB.user|https://github.com/zonemaster/zonemaster-backend/blob/master/docs/Configuration.md#db>
```
`#mysql-section` or `#user-1`
`#db-section` or `#user`

pnax commented on lib/Zonemaster/Backend/Config.pm:
```diff
@@ -58,22 +72,130 @@ sub BackendDBType {
+
+=head2 POSTGRESQL_database
+
+Returns the L<POSTGRESQL.database|https://github.com/zonemaster/zonemaster-backend/blob/master/docs/Configuration.md#postgresql>
+property from the loaded config, or the L<DB.database_name|https://github.com/zonemaster/zonemaster-backend/blob/master/docs/Configuration.md#db>
```
`#postgresql-section` or `#database-1`
`#db-section` or `#database_name`

pnax commented on lib/Zonemaster/Backend/Config.pm:
```diff
@@ -58,22 +72,130 @@ sub BackendDBType {
-    return $self->{cfg}->val( 'DB', 'database_name' );
+=head2 POSTGRESQL_host
+
+Returns the L<POSTGRESQL.host|https://github.com/zonemaster/zonemaster-backend/blob/master/docs/Configuration.md#postgresql>
+property from the loaded config, or the L<DB.database_host|https://github.com/zonemaster/zonemaster-backend/blob/master/docs/Configuration.md#db>
```
`#postgresql-section` or `#host-1`
`#db-section` or `#database_host`

pnax commented on lib/Zonemaster/Backend/Config.pm:
```diff
@@ -58,22 +72,130 @@ sub BackendDBType {
+
+=head2 POSTGRESQL_password
+
+Returns the L<POSTGRESQL.password|https://github.com/zonemaster/zonemaster-backend/blob/master/docs/Configuration.md#postgresql>
+property from the loaded config, or the L<DB.password|https://github.com/zonemaster/zonemaster-backend/blob/master/docs/Configuration.md#db>
```
`#postgresql-section` or `#password-1`
`#db-section` or `#password`

pnax commented on lib/Zonemaster/Backend/Config.pm:
```diff
@@ -58,22 +72,130 @@ sub BackendDBType {
+
+=head2 POSTGRESQL_user
+
+Returns the L<POSTGRESQL.user|https://github.com/zonemaster/zonemaster-backend/blob/master/docs/Configuration.md#postgresql>
+property from the loaded config, or the L<DB.user|https://github.com/zonemaster/zonemaster-backend/blob/master/docs/Configuration.md#db>
```
`#postgresql-section` or `#user-2`
`#db-section` or `#user`

pnax commented on lib/Zonemaster/Backend/Config.pm:
```diff
@@ -58,22 +72,130 @@ sub BackendDBType {
+
+=head2 SQLITE_file
+
+Returns the L<SQLITE.file|https://github.com/zonemaster/zonemaster-backend/blob/master/docs/Configuration.md#sqlite>
+property from the loaded config, or the L<DB.database_name|https://github.com/zonemaster/zonemaster-backend/blob/master/docs/Configuration.md#db>
```
`#postgresql-section` or `#file`
`#db-section` or `#database_name`

pnax reviewed (commented):
This looks really good !

mattias-p commented:
I've rebased this PR onto the current develop branch and updated the testing instructions.

It kindof sucks that the rpcapi logs only go the systemd journal. I haven't checked where they go on FreeBSD. Maybe /var/log/messages? The reason is that the rpcapi currently doesn't integrate well with starman when it comes to logging. If you like I can update this PR to also include an improvement to the log integration, or if you'd prefer to have this merged first and have the log integration fixed in a follow-up PR, that's okay with me too.

matsduf commented:
> It kindof sucks that the rpcapi logs only go the systemd journal. I haven't checked where they go on FreeBSD. Maybe /var/log/messages? 

No, there is no trace of those log messages in FreeBSD as far as I can see.

> The reason is that the rpcapi currently doesn't integrate well with starman when it comes to logging. If you like I can update this PR to also include an improvement to the log integration, or if you'd prefer to have this merged first and have the log integration fixed in a follow-up PR, that's okay with me too.

I prefer to have logging updated in a separate PR. It is better to have several, but smaller PRs when possible.

mattias-p commented:
I created #736 to addresses the mishandling of the log messages.

@matsduf, @pnax Please, review.

matsduf commented on docs/Installation.md:
````diff
@@ -591,7 +583,6 @@ path:
 
 ```sh
 sed -i '' '/[[:<:]]engine[[:>:]]/ s/=.*/= SQLite/' /usr/local/etc/zonemaster/backend_config.ini
-sed -i '' '/[[:<:]]database_name[[:>:]]/ s:=.*:= /var/db/zonemaster/db.sqlite:' /usr/local/etc/zonemaster/backend_config.ini
````
In FreeBSD, the location of the SQLite database file should be /var/db/zonemaster, not /var/lib/zonemaster.

mattias-p replied:
Good catch! I'll add that line back.

matsduf commented on docs/Installation.md:
````diff
@@ -591,7 +583,7 @@ path:
 ```sh
 sed -i '' '/[[:<:]]engine[[:>:]]/ s/=.*/= SQLite/' /usr/local/etc/zonemaster/backend_config.ini
-sed -i '' '/[[:<:]]database_name[[:>:]]/ s:=.*:= /var/db/zonemaster/db.sqlite:' /usr/local/etc/zonemaster/backend_config.ini
+sed -i '' '/[[:<:]]file[[:>:]]/ s:=.*:= /var/db/zonemaster/db.sqlite:' /usr/local/etc/zonemaster/backend_config.ini
````
For this replacement to work, there cannot be any other key "file", which is not unlikely since that is a generic term. Maybe "databasefile" would be better?

mattias-p replied:
Apparently SQLite can use more than one file. https://sqlite.org/fileformat.html
I'll rename it to database_file.

matsduf commented on docs/Configuration.md:
```diff
@@ -34,32 +33,103 @@ SQLite            | `SQLite`
 
-The name of the database to use, except for SQLite database engine where
-it holds the full path to the SQLite database file.
+**Deprecated.** Use [MYSQL.database], [POSTGRESQL.database] or [SQLite.file] instead.
```
"SQLite.file" => "SQLite.database_file"

mattias-p replied:
My God I missed a lot of them.

matsduf commented on docs/Configuration.md:
```diff
@@ -34,32 +33,103 @@ SQLite            | `SQLite`
-it holds the full path to the SQLite database file.
+**Deprecated.** Use [MYSQL.database], [POSTGRESQL.database] or [SQLite.file] instead.
+
+The [MYSQL.database], [POSTGRESQL.database], [SQLite.file] properties take precedence
```
"SQLite.file" => "SQLite.database_file"

matsduf commented on docs/Configuration.md:
```diff
@@ -208,15 +278,28 @@ zero minutes, then the default value (600 seconds) is used.
 [Profile JSON files]:                 https://github.com/zonemaster/zonemaster-engine/blob/master/docs/Profiles.md
 [Profile names]:                      API.md#profile-name
 [Profiles]:                           Architecture.md#profile
+[SQLite.file]:                        #file
```
```
[SQLite.database_file]:     #database_file
```

matsduf commented on share/backend_config.ini:
```diff
@@ -3,12 +3,23 @@
+database = zonemaster
+
+[SQLITE]
+file = /var/lib/zonemaster/db.sqlite
```
"file" => "database_file"

matsduf commented on share/travis_sqlite_backend_config.ini:
```diff
@@ -1,14 +1,11 @@
 #seconds
 
+[SQLITE]
+file=/tmp/zonemaster
```
"file" => "database_file"

matsduf commented on lib/Zonemaster/Backend/DB/SQLite.pm:
```diff
@@ -31,18 +31,17 @@ sub BUILD {
     if ( !defined $self->dbh ) {
-        my $connection_string = $self->config->DB_connection_string( 'sqlite' );
-        $log->debug( "Connection string: " . $connection_string );
+        my $file = $self->config->SQLITE_file;
```
```
my $file = $self->config->SQLITE_database_file;
```

matsduf reviewed (approved):
I assume that all "file" <=> "database_file" have been found.

mattias-p commented:
@pnax Could you review this again?

pnax commented:
Yes, sorry, I'll do it.

pnax commented on share/zm-testagent.lsb:
```diff
@@ -20,6 +20,8 @@ PIDFILE=${ZM_BACKEND_PIDFILE:-/var/run/zonemaster/zm-testagent.pid}
 USER=${ZM_BACKEND_USER:-zonemaster}
 GROUP=${ZM_BACKEND_GROUP:-zonemaster}
 
+#export ZM_BACKEND_TESTAGENT_LOGLEVEL='warning'  # Set this variable to override the default log level
```
`ZM_BACKEND_TESTAGENT_LOGLEVEL` is not used in `zonemaster_backend_testagent`. Maybe an omission.

Furthermore this might be in conflict with the option `--loglevel`.

mattias-p replied:
Good spotting! I made an update to make sure the log level value is propagated.

pnax commented on share/zm_testagent-bsd:
```diff
@@ -16,6 +16,7 @@ load_rc_config $name
 : ${zm_testagent_pidfile="/var/run/zonemaster/${name}.pid"}
 
 export ZONEMASTER_BACKEND_CONFIG_FILE="/usr/local/etc/zonemaster/backend_config.ini"
+#export ZM_BACKEND_TESTAGENT_LOGLEVEL='warning'  # Set this variable to override the default log level
```
same remark on `ZM_BACKEND_TESTAGENT_LOGLEVEL`.

mattias-p replied:
I fixed this the same way.

pnax reviewed (commented):
See my remark on `ZM_BACKEND_TESTAGENT_LOGLEVEL`.

Otherwise, I tested it and I was able :

* to launch the daemons with a new and an old config file
* with the old config file, I can see the deprecation warnings
* with the new config file, I am able to create a test on a domain

About the logs:

* it seems that the `Opening SQLite` message is issued twice (maybe because we test the connection first, then we open it)
* trying with PostgreSQL, each time I send a query to the API, there is the `Connecting to PostgreSQL` message appended (it looks like we reconnect to the database each time, is that so?)

mattias-p commented:
> About the logs (level NOTICE) :
>
> * it seems that the Opening SQLite message is issued twice (maybe because we test the connection first, then we open it)
> * trying with PostgreSQL, each time I send a query to the API, there is the Connecting to PostgreSQL message appended (it looks like we reconnect to the database each time, is that so?)

Those are good observations! Could you create issues for those? I'd like to limit the scope of this PR. It's already grown way beyond my initial intentions.

pnax commented on share/zm-testagent.lsb:
```diff
@@ -20,9 +20,9 @@ PIDFILE=${ZM_BACKEND_PIDFILE:-/var/run/zonemaster/zm-testagent.pid}
+#export ZM_BACKEND_TESTAGENT_LOGLEVEL='info'  # Set this variable to override the default log level
 
-testagent_args="--logfile=$LOGFILE --outfile=$OUTFILE --pidfile=$PIDFILE --user=$USER --group=$GROUP"
+testagent_args="--loglevel=$ZM_BACKEND_TESTAGENT_LOGLEVEL --logfile=$LOGFILE --outfile=$OUTFILE --pidfile=$PIDFILE --user=$USER --group=$GROUP"
```
If you pass it directly to the arguments, shouldn't the variable be exported then ?

mattias-p replied:
Since the variable is only used locally it doesn't need to be exported. And since the default value is defined in zonemaster_backend_testagent I don't want to define it again in the init scripts - for the sake of having a single source of truth.

But you're completely right that there is something fishy going on here! I'm setting --loglevel to the empty string when I ZM_BACKEND_TESTAGENT_LOGLEVEL is undefined. I'll fix that.

pnax commented on share/zm_testagent-bsd:
```diff
@@ -16,10 +16,10 @@ load_rc_config $name
 
 export ZONEMASTER_BACKEND_CONFIG_FILE="/usr/local/etc/zonemaster/backend_config.ini"
-#export ZM_BACKEND_TESTAGENT_LOGLEVEL='warning'  # Set this variable to override the default log level
+#export ZM_BACKEND_TESTAGENT_LOGLEVEL='info'  # Set this variable to override the default log level
```
same

pnax reviewed (commented):
Sorry to be punctilious, shouldn't the variable `ZM_BACKEND_TESTAGENT_LOGLEVEL` be set then ?

pnax commented:
> Those are good observations! Could you create issues for those? I'd like to limit the scope of this PR. It's already grown way beyond my initial intentions.

Having a quick look on this, it seems to be the expected behavior. We can keep it like this for now. (Sorry to overload the PR with my remarks)

mattias-p commented:
> Sorry to overload the PR with my remarks

In that case I'm sorry for making so many silly mistakes.

pnax reviewed (approved):
This looks good, thanks :)

mattias-p commented:
## Release testing for v2021.1

I've verified this test procedure on CentOS 7 and CentOS 8.

(However to get it to work I had to replace the `service` commands with `systemctl` dittos.)
